cli/build/main.rs

```diff
@@ -166,71 +136,45 @@ fn do_main() -> Result<(), Error> {
-			assert!(stack_size <= 1024*1024);
-			let (new_module, new_stack_top) = shrink_unknown_stack(module, 1024 * 1024 - stack_size);
-			module = new_module;
-			let mut stack_top_page = new_stack_top / 65536;
-			if new_stack_top % 65536 > 0 { stack_top_page += 1 };
-			module = externalize_mem(module, Some(stack_top_page), 16);
-		} else {
-			module = externalize_mem(module, None, 16);
-		}
-	}
-
-	if let Some(runtime_type) = matches.value_of("runtime_type") {
-		let runtime_type: &[u8] = runtime_type.as_bytes();
-		if runtime_type.len() != 4 {
+	let (runtime_type, runtime_version) = if let (Some(runtime_type), Some(runtime_version)) = (matches.value_of("runtime_type"), matches.value_of("runtime_version")) {
+		let ty: &[u8] = runtime_type.as_bytes();
+		if ty.len() != 4 {
 			panic!("--runtime-type should be equal to 4 bytes");
 		}
-		let runtime_version: u32 = matches.value_of("runtime_version").unwrap_or("1").parse()
+		let version: u32 = runtime_version.parse()
 			.expect("--runtime-version should be a positive integer");
-		module = utils::inject_runtime_type(module, &runtime_type, runtime_version);
-	}
-
-	let mut ctor_module = module.clone();
+		(Some(ty), Some(version))
+	} else {
+		(None, None)
+	};
 
-	let mut public_api_entries = matches.value_of("public_api")
+	let public_api_entries = matches.value_of("public_api")
 		.map(|val| val.split(",").collect())
 		.unwrap_or(Vec::new());
-	public_api_entries.push(CALL_SYMBOL);
-	if !matches.is_present("skip_optimization") {
-		utils::optimize(
-			&mut module,
-			public_api_entries,
-		)?;
-	}
+
+	let (module, ctor_module) = build(
+		module,
+		true,
+		source_input.target(),
+		runtime_type, runtime_version,
+		public_api_entries,
+		matches.is_present("enforce_stack_adjustment"),
+		matches.value_of("shrink_stack").unwrap_or_else(|| "49152").parse()
+			.expect("New stack size is not valid u32"),
+		matches.is_present("skip_optimization"),
+	).map_err(Error::Build)?;
 
 	if let Some(save_raw_path) = matches.value_of("save_raw") {
 		parity_wasm::serialize_to_file(save_raw_path, module.clone()).map_err(Error::Encoding)?;
 	}
 
-	let raw_module = parity_wasm::serialize(module).map_err(Error::Encoding)?;
-
-	// If module has an exported function with name=CREATE_SYMBOL
-	// build will pack the module (raw_module) into this funciton and export as CALL_SYMBOL.
-	// Otherwise it will just save an optimised raw_module
-	if has_ctor(&ctor_module) {
-		if !matches.is_present("skip_optimization") {
-			utils::optimize(&mut ctor_module, vec![CREATE_SYMBOL])?;
-		}
-		let ctor_module = utils::pack_instance(raw_module, ctor_module)?;
-		parity_wasm::serialize_to_file(&path, ctor_module).map_err(Error::Encoding)?;
-	} else {
-		let mut file = fs::File::create(&path)?;
-		file.write_all(&raw_module)?;
-	}
-
+	parity_wasm::serialize_to_file(
+		&path,
+		ctor_module.expect("ctor_module can't be None, because 'constructor' argument is set to true in build")
```

Comment:
there should be a comma after the last argument when each is on a dedicated line